packages/router/src/router.tsx

```diff
@@ -1,34 +1,116 @@
 // The guts of the router implementation.
 
-import React, { ReactChild, ReactElement } from 'react'
+import React, { ReactChild } from 'react'
 
-import { useAuth as useAuthHook } from '@redwoodjs/auth'
+// TODO: Private routes!
+// import { useAuth as useAuthHook } from '@redwoodjs/auth'
 
 import {
-  Location,
   parseSearch,
   replaceParams,
   matchPath,
-  mapNamedRoutes,
   PageLoader,
   Redirect,
-  LocationContextType,
-  ParamType,
+  useLocation,
+  validatePath,
+  LocationProvider,
 } from './internal'
+import {
+  RouterContextProvider,
+  RouterState,
+  useRouterState,
+} from './router-context'
 import { SplashPage } from './splash-page'
 
+/**
+ * A more specific interface will be generated by
+ * babel-plugin-redwood-routes-auto-loader when a redwood project is built
+ *
+ * Example:
+ * interface AvailableRoutes {
+ *   home: () => "/"
+ *   test: () => "/test"
+ * }
+ */
+interface AvailableRoutes {
+  [key: string]: (args?: Record<string, string>) => string
+}
+
+const namedRoutes: AvailableRoutes = {}
+
 interface RouteProps {
-  page: Spec | React.ComponentType<unknown> | ((props: any) => JSX.Element)
   path?: string
+  page: Spec | React.ComponentType<unknown> | ((props: any) => JSX.Element)
   name?: string
   notfound?: boolean
   redirect?: string
   prerender?: boolean
   whileLoading?: () => ReactChild | null
 }
 
-const Route: React.VFC<RouteProps> = () => {
-  return null
+const Route: React.VFC<RouteProps> = ({
+  path,
+  page,
+  name,
+  redirect,
+  notfound,
+  whileLoading,
+}) => {
+  const location = useLocation()
+  const routerState = useRouterState()
+
+  if (notfound) {
+    // TODO: How should we handle `notfound`?
```

Comment:
Please help me figure something out here :)

Reply:
Redirect to /404 ?

Reply:
Perhaps, but the real question is, how do we know no route was matched? And when do we detect and handle this?

Reply:
Is a `NotFound` allowed be be contained within a `Set`? 

It strikes me that `NotFound` should be a special route - just one and at root (ie, not in a Set).

Reply:
Yeah, it would only make sense to have one of those. And at the root, like you said. There is currently nothing in place to check this. Maybe something could be added to the structure package.